pkg/steps/multi_stage.go

```diff
@@ -0,0 +1,202 @@
+		name:      testConfig.As,
+		config:    config,
+		podClient: podClient,
+		jobSpec:   jobSpec,
+		pre:       testConfig.MultiStageTestConfiguration.Pre,
+		test:      testConfig.MultiStageTestConfiguration.Test,
+		post:      testConfig.MultiStageTestConfiguration.Post,
+	}
+}
+
+func (s *multiStageTestStep) Inputs(ctx context.Context, dry bool) (api.InputDefinition, error) {
+	return nil, nil
+}
+
+func (s *multiStageTestStep) Run(ctx context.Context, dry bool) error {
+	if dry {
+		return nil
+	}
+	var errs []error
+	if err := s.runSteps(ctx, s.pre, true); err != nil {
+		errs = append(errs, fmt.Errorf("%q pre steps failed: %v", s.name, err))
+	} else if err := s.runSteps(ctx, s.test, true); err != nil {
+		errs = append(errs, fmt.Errorf("%q test steps failed: %v", s.name, err))
+	}
+	if err := s.runSteps(ctx, s.post, false); err != nil {
+		errs = append(errs, fmt.Errorf("%q post steps failed: %v", s.name, err))
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (s *multiStageTestStep) Done() (bool, error) { return false, nil }
+func (s *multiStageTestStep) Name() string        { return s.name }
+func (s *multiStageTestStep) Description() string {
+	return fmt.Sprintf("Run test %s", s.name)
+}
+
+func (s *multiStageTestStep) Requires() (ret []api.StepLink) {
+	var needsImages, needsRelease bool
+	for _, step := range append(append(s.pre, s.test...), s.post...) {
+		if s.config.IsPipelineImage(step.Image) {
+			ret = append(ret, api.InternalImageLink(api.PipelineImageStreamTagReference(step.Image)))
+		} else if s.config.BuildsImage(step.Image) {
+			needsImages = true
+		} else {
+			needsRelease = true
+		}
+	}
+	if needsImages {
+		ret = append(ret, api.ImagesReadyLink())
+	}
+	if needsRelease {
+		ret = append(ret, api.ReleaseImagesLink())
+	}
+	return
+}
+
+func (s *multiStageTestStep) Creates() []api.StepLink { return nil }
+func (s *multiStageTestStep) Provides() (api.ParameterMap, api.StepLink) {
+	return nil, nil
+}
+func (s *multiStageTestStep) SubTests() []*junit.TestCase { return s.subTests }
+
+func (s multiStageTestStep) runSteps(ctx context.Context, steps []api.TestStep, shortCircuit bool) error {
+	pods, err := s.genPods(steps)
+	if err != nil {
+		return err
+	}
+	return s.runPods(ctx, pods, shortCircuit)
+}
+
+func (s multiStageTestStep) genPods(steps []api.TestStep) ([]coreapi.Pod, error) {
+	var ret []coreapi.Pod
+	var errs []error
+	for _, step := range steps {
+		image := step.Image
+		if s.config.IsPipelineImage(step.Image) {
+			image = api.PipelineImageStream + ":" + step.Image
+		}
+		resources, err := resourcesFor(step.Resources)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		name := s.name + "-" + step.Name
+		pod, err := generateBasePod(s.jobSpec, name, step.Name, []string{"/bin/bash", "-c", "#!/bin/bash\nset -eu\n" + step.Commands}, image, resources)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		pod.Labels[multiStageTestLabel] = s.name
```

Comment:
since you are doing this, can you add also the [org/repo/branch](https://github.com/openshift/ci-tools/blob/master/pkg/steps/template.go#L35-L37) labels?

Reply:
That should happen automatically after https://github.com/openshift/ci-tools/pull/92 merges.